cranelift-codegen/meta/src/shared/legalize.rs

```diff
@@ -272,6 +272,22 @@ pub(crate) fn define(insts: &InstructionGroup, imm: &Immediates) -> TransformGro
         ],
     );
 
+    // TODO(ryzokuken): explore the perf diff w/ x86_umulx and consider have a
+    // separate legalization for x86.
+    narrow.legalize(
+        def!(a = imul.I64(x, y)),
```

Comment:
Could you please also implement this and further i64 -> i32 narrowing legalizations for i128?

Reply:
I could legalize `imul.ty` in terms of `imul.ty_half` which is good, but I guess I'll need to legalize `umulhi` for I64 for that too?

Reply:
For x86_64 it is already implemented: https://github.com/CraneStation/cranelift/blob/6ed06a1e8e428aedb7adfd825b131abb9a7a3d15/cranelift-codegen/meta/src/isa/x86/encodings.rs#L725

32bit support

Reply:
so `umulhi` is defined for x86_64. I guess it's okay to add the legalization since it's not defined for any other ISA anyway...